### PR TITLE
feat(invity): sell crypto feature

### PR DESCRIPTION
**Pull request by martinboehm**

### CODE REVIEW

File: packages/suite/src/components/suite/modals/confirm/CoinmarketSellTerms/index.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+`;
+
+export type Props = {
+    decision: Deferred<boolean>;
+    onCancel: () => void;
+    provider?: string;
+};
+
+const CoinmarketSellTerms = ({ decision, onCancel, provider }: Props) => {
+    const [isChecked, setIsChecked] = useState<boolean>(false);
```
Comment (szymonlesisz) [inline packages/suite/src/components/suite/modals/confirm/CoinmarketSellTerms/index.tsx]:
```suggestion
    const [isChecked, setIsChecked] = useState(false);
```
nitpick: type is obvious

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+const Arrow = styled.div`
+    display: flex;
+    align-items: center;
+    padding: 0 11px;
+`;
+
+const SellTransaction = ({ trade, providers, account }: Props) => {
+    const theme = useTheme();
+    const { goto } = useActions({ goto: routerActions.goto });
+    const { saveTransactionDetailId } = useActions({
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx]:
```suggestion
    const { saveTransactionDetailId, goto } = useActions({
```
merge useAction hook

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            accountIndex: account.index,
+            accountType: account.accountType,
+        });
+    };
+
+    return (
+        <Wrapper>
+            <Column>
+                <Row>
+                    {/* TODO FIX THIS LOGO */}
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx]: tracking TODO

File: packages/suite/src/actions/wallet/coinmarketSellActions.ts [hunk 4]
```diff
@@ -19,9 +50,22 @@ export const loadSellInfo = async (): Promise<SellInfo> => {
         sellList.providers.forEach(e => (providerInfos[e.name] = e));
     }
 
+    const tradedFiatCurrencies: string[] = [];
+    const tradedCoins: string[] = [];
+    sellList?.providers.forEach(p => {
+        if (p.tradedFiatCurrencies) {
+            tradedFiatCurrencies.push(...p.tradedFiatCurrencies.map(c => c.toLowerCase()));
+        }
+        tradedCoins.push(...p.tradedCoins.map(c => c.toLowerCase()));
```
Comment (szymonlesisz) [inline packages/suite/src/actions/wallet/coinmarketSellActions.ts]: discuss

File: packages/suite/src/views/wallet/coinmarket/sell/offers/index.tsx [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+const Wrapper = styled.div`
+    display: flex;
+    width: 100%;
+    flex-direction: column;
+`;
+
+const OffersIndexLoaded = (props: Props) => {
+    const { selectedAccount } = props;
+    const coinmarketOffersValues = useOffers({ ...props, selectedAccount });
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/offers/index.tsx]:
```suggestion
    const coinmarketOffersValues = useOffers(props);
```
is spreading necessary?

File: packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SelectBankAccount/index.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+                    </RegisterAnother>
+                </Row>
+            </CardContent>
+            <ButtonWrapper>
+                <Button
+                    isLoading={callInProgress}
+                    onClick={() => {
+                        if (bankAccount) confirmTrade(bankAccount);
+                    }}
+                    isDisabled={callInProgress}
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SelectBankAccount/index.tsx]:
```suggestion
                    isDisabled={callInProgress || !bankAccount}
```

File: packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/List/Quote/index.tsx [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+
+const StyledQuestionTooltip = styled(QuestionTooltip)`
+    padding-left: 4px;
+    color: ${props => props.theme.TYPE_LIGHT_GREY};
+`;
+
+export function getQuoteError(quote: SellFiatTrade, amountInCrypto: boolean) {
+    if (quote.error) {
+        if (amountInCrypto) {
+            if (quote.minCrypto && Number(quote.cryptoStringAmount) < quote.minCrypto) {
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/List/Quote/index.tsx]: `Number(quote.cryptoStringAmount)` conversion is used multiple times in this code block, could be assigned as const?

File: packages/suite/src/views/wallet/coinmarket/sell/index.tsx [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import { CoinmarketLayout, WalletLayout } from '@wallet-components';
+import { useSelector } from '@suite-hooks';
+import { Props } from '@wallet-types/coinmarketSellForm';
+import SellForm from './components/SellForm';
+import { useCoinmarketSellForm, SellFormContext } from '@wallet-hooks/useCoinmarketSellForm';
+
+const CoinmarketSellLoaded = (props: Props) => {
+    const { selectedAccount } = props;
+    const coinmarketSellContextValues = useCoinmarketSellForm({
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/index.tsx]:
```suggestion
    const coinmarketSellContextValues = useCoinmarketSellForm(props)
```
unnecessary spread?

File: packages/suite/src/views/wallet/coinmarket/sell/detail/index.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+const Wrapper = styled.div`
+    display: flex;
+    width: 100%;
+    flex-direction: column;
+`;
+
+const DetailIndexLoaded = (props: Props) => {
+    const { selectedAccount } = props;
+    const coinmarketSellContextValues = useCoinmarketSellDetail({
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/detail/index.tsx]: spread :)

File: packages/suite/src/views/wallet/coinmarket/sell/components/SellForm/Inputs/FiatInput/index.tsx [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+                                  value: quotesRequest.fiatCurrency.toUpperCase(),
+                              }
+                            : defaultCurrency
+                    }
+                    render={({ onChange, value }) => {
+                        return (
+                            <Select
+                                options={FIAT.currencies
+                                    .filter(c => sellInfo?.supportedFiatCurrencies.has(c))
+                                    .map((currency: string) => buildOption(currency))}
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/components/SellForm/Inputs/FiatInput/index.tsx]:
```suggestion
                                    .map(buildOption)
```

File: packages/suite/src/views/wallet/coinmarket/buy/detail/components/WaitingForPayment/index.tsx [hunk 11]
```diff
@@ -54,15 +54,15 @@ interface Props {
...
+        submitRequestForm: coinmarketCommonActions.submitRequestForm,
     });
 
     const goToPayment = async () => {
         setIsWorking(true);
         const returnUrl = await createTxLink(trade, account);
         invityAPI.getBuyTradeForm({ trade, returnUrl }).then(response => {
             if (response) {
-                submitRequestForm(response);
+                submitRequestForm(response.form);
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/buy/detail/components/WaitingForPayment/index.tsx]: nitpick: any reason to mix `async/await` with `then` syntax? (i didn't noticed that earlier)

File: packages/suite/src/utils/wallet/coinmarket/sellUtils.ts [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+// loop through quotes and if all quotes are either with error below minimum or over maximum, return the limits
+export function getAmountLimits(
+    request: SellFiatTradeQuoteRequest,
+    quotes: SellFiatTrade[],
+): AmountLimits | undefined {
+    let minAmount: number | undefined;
+    let maxAmount: number | undefined;
+    // eslint-disable-next-line no-restricted-syntax
+    for (const quote of quotes) {
```
Comment (szymonlesisz) [inline packages/suite/src/utils/wallet/coinmarket/sellUtils.ts]: eslint discuss

File: packages/suite/src/hooks/wallet/useCoinmarketSellOffers.ts [hunk 13]
```diff
@@ -0,0 +1,301 @@
...
+    const sendTransaction = async () => {
+        if (
+            selectedQuote &&
+            selectedQuote.orderId &&
+            selectedQuote.destinationAddress &&
+            selectedQuote.cryptoStringAmount &&
+            composed
+        ) {
+            // prepare the fee levels, set custom values from composed
+            // WORKAROUND: sendFormEthereumActions and sendFormRippleActions use form outputs instead of composed transaction data
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketSellOffers.ts]: discuss

File: packages/suite/src/components/suite/modals/confirm/CoinmarketSellTerms/index.tsx (hunk 1 above)
Comment (martinboehm) [reply packages/suite/src/components/suite/modals/confirm/CoinmarketSellTerms/index.tsx]: done

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx (hunk 2 above)
Comment (martinboehm) [reply packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx]: done

File: packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx (hunk 3 above)
Comment (martinboehm) [reply packages/suite/src/components/wallet/CoinmarketLayout/components/AccountTransactions/components/SellTransaction/index.tsx]: done

File: packages/suite/src/actions/wallet/coinmarketSellActions.ts (hunk 4 above)
Comment (martinboehm) [reply packages/suite/src/actions/wallet/coinmarketSellActions.ts]: done

File: packages/suite/src/views/wallet/coinmarket/sell/offers/index.tsx (hunk 5 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/offers/index.tsx]: done

File: packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SelectBankAccount/index.tsx (hunk 6 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SelectBankAccount/index.tsx]: done

File: packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/List/Quote/index.tsx (hunk 7 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/List/Quote/index.tsx]: done

File: packages/suite/src/views/wallet/coinmarket/sell/index.tsx (hunk 8 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/index.tsx]: done

File: packages/suite/src/views/wallet/coinmarket/sell/detail/index.tsx (hunk 9 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/detail/index.tsx]: done

File: packages/suite/src/views/wallet/coinmarket/buy/detail/components/WaitingForPayment/index.tsx (hunk 11 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/buy/detail/components/WaitingForPayment/index.tsx]: done

File: packages/suite/src/utils/wallet/coinmarket/sellUtils.ts (hunk 12 above)
Comment (martinboehm) [reply packages/suite/src/utils/wallet/coinmarket/sellUtils.ts]: done

File: packages/suite/src/hooks/wallet/useCoinmarketSellOffers.ts (hunk 13 above)
Comment (martinboehm) [reply packages/suite/src/hooks/wallet/useCoinmarketSellOffers.ts]: Refactored to useCoinmarketRecomposeAndSign hook in sell and exchange crypto.

**Review by szymonlesisz — APPROVED** [review]
looks good, didn't test it tho

File: packages/suite/src/components/wallet/CoinmarketPaymentType/index.tsx [hunk 14]
```diff
@@ -53,8 +55,13 @@ const CoinmarketPaymentType = ({ method }: Props) => (
                         />
                     </Bg>
                 </IconWrapper>
-                {/* temporary solution - payment mehtod name will be returned by API server to be independent on translations */}
-                <Translation id={`TR_PAYMENT_METHOD_${method.toUpperCase()}` as any} />
+                <div>
```
Comment (szymonlesisz) [inline packages/suite/src/components/wallet/CoinmarketPaymentType/index.tsx:58]:
```suggestion
                <>
```
`div` maybe replaced by react fragment

File: packages/suite/src/hooks/wallet/useCoinmarketSellForm.ts [hunk 15]
```diff
@@ -0,0 +1,312 @@
...
+    );
+
+    // watch change in fiat amount and recalculate fees on change
+    const onFiatAmountChange = useCallback(
+        (amount: string) => {
+            if (!amount) return;
+            setValue(CRYPTO_INPUT, '');
+            setValue('setMaxOutputId', undefined);
+            clearErrors(CRYPTO_INPUT);
+            const currency: { value: string; label: string } | undefined = getValues(
```
Comment (szymonlesisz) [inline packages/suite/src/hooks/wallet/useCoinmarketSellForm.ts]:
```suggestion
            const currency:  typeof defaultCurrency | undefined = getValues(
```

File: packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SendTransaction/index.tsx [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+const SendTransactionComponent = () => {
+    const {
+        account,
+        callInProgress,
+        selectedQuote,
+        sellInfo,
+        sendTransaction,
+    } = useCoinmarketSellOffersContext();
+    if (!selectedQuote) return null;
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SendTransaction/index.tsx]:
```suggestion
    if (!selectedQuote || !selectedQuote.exchange) return null;
```

File: packages/suite/src/views/wallet/coinmarket/sell/detail/components/PaymentPending/index.tsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+const PaymentConverting = ({ supportUrl }: Props) => (
+    <Wrapper>
+        <Loader />
+        <Title>
+            <Translation id="TR_SELL_DETAIL_PENDING_TITLE" />
+        </Title>
+        {supportUrl && (
+            <Link href={supportUrl} target="_blank">
```
Comment (szymonlesisz) [inline packages/suite/src/views/wallet/coinmarket/sell/detail/components/PaymentPending/index.tsx]: use @trezor/components/Link 
related to https://github.com/trezor/trezor-suite/issues/3633

File: packages/suite/src/components/wallet/CoinmarketPaymentType/index.tsx (hunk 14 above)
Comment (martinboehm) [reply packages/suite/src/components/wallet/CoinmarketPaymentType/index.tsx:58]: Not done, the `div` has visual function

File: packages/suite/src/hooks/wallet/useCoinmarketSellForm.ts (hunk 15 above)
Comment (martinboehm) [reply packages/suite/src/hooks/wallet/useCoinmarketSellForm.ts]: Done

File: packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SendTransaction/index.tsx (hunk 16 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/offers/Offers/SelectedOffer/components/SendTransaction/index.tsx]: Done

File: packages/suite/src/views/wallet/coinmarket/sell/detail/components/PaymentPending/index.tsx (hunk 17 above)
Comment (martinboehm) [reply packages/suite/src/views/wallet/coinmarket/sell/detail/components/PaymentPending/index.tsx]: Done (also in all similar cases in buy/exchange/sell functionality)

**Review by szymonlesisz — APPROVED** [review]

**Comment by sorooris** [comment]
**QA NOK**

We are unable to complete the Sell flow in any of the provided testing builds:

- None of the combinations of production and staging APIs with BTC and TESTNET  alllows to complete the flow

1. Production API: 

- BTC - SelectedOffer__StyledCard is missing content
- ETH - Currency limit exceeded
- TESTNET - no providers

2. Staging API:
- BTC, ETH - Currency limit exceeded
- TESTNET - same as prod, no providers

**Setup**
Suite: app, web 21.5.1 (4d09d884)
Invity API server: production (https://exchange.trezor.io/), staging (http://staging-exchange.invity.io/)
Browser: N/A, Firefox 86.0.1
OS: nixOS 20.09, macOS Big Sur 11.3.1
Device: T1 1.9.0

**Screenshots**:
Production API - BTC - SelectedOffer__StyledCard missing content
image
TESTNET - with both prod and staging API
image
Staging API - BTC
image

**Comment by martinboehm** [comment]
Retested together with QA, the functionality is basically OK.
The majority of confusion comes from using shared Trezor and shared BTC Direct account for testing, which does not correspond to the normal user flow.
Identified a minor bug (not disabled button) in one edge case. While the bug must be fixed, it is not a release stopper.

**Comment by sorooris** [comment]
ack, thanks!

**Pull request merged**